app/shared/Shared/Services/AzureSearchEmbedService.cs

```diff
@@ -315,18 +325,97 @@ private async Task UploadCorpusAsync(string corpusBlobName, string text)
+                    string Content,
+                    string SourcePage,
+                    string SourceFile,
+                    string? Category = null,
+                    Tokenizer? tokenizer = null)
+    {
+        tokenizer ??= await Tiktoken.CreateByModelNameAsync(modelEncodingName); 
+        if (tokenizer.CountTokens(Content) <= maxTokensPerSection)
+        {
+            yield return new Section(
+                Id: Id,
+                Content: Content,
+                SourcePage: SourcePage,
+                SourceFile: SourceFile,
+                Category: Category);
+        } else
+        {
+            int start = Content.Length / 2;
+            int pos = 0;
+            int boundary = Content.Length / 3;
+            int splitPosition = -1;
+
+            while (start - pos > boundary)
+            {
+                if (s_sentenceEndings.Contains(Content[start - pos]))
+                {
+                    splitPosition = start - pos;
+                    break;
+                }
+                else if (s_sentenceEndings.Contains(Content[start + pos]))
+                {
+                    splitPosition = start + pos;
+                    break;
+                }
+                else
+                {
+                    pos += 1;
+                }
+            }
+
+            string firstHalf, secondHalf;
+
+            if (splitPosition > 0)
+            {
+                firstHalf = Content[..(splitPosition + 1)];
+                secondHalf = Content[(splitPosition + 1)..];
+            }
+            else
+            {
+                // Split page in half and call function again
+                // Overlap first and second halves by DEFAULT_OVERLAP_PERCENT%
+                int middle = Content.Length / 2;
+                int overlapChars = (int)(Content.Length * (DefaultOverlapPercent / 100.0));
+                firstHalf = Content[..(middle + overlapChars)];
+                secondHalf = Content[(middle - overlapChars)..];
+            }
+
+            await foreach(var section in SplitSectionByTokenLengthAsync(Id, firstHalf, SourcePage, SourceFile, Category, tokenizer))
+            {
+                yield return section;
+            }
+            await foreach (var section in SplitSectionByTokenLengthAsync(Id, secondHalf, SourcePage, SourceFile, Category, tokenizer))
+            {
+                yield return section;
+            }
+        }
+        
+    }
+
+    public async IAsyncEnumerable<Section> CreateSectionsAsync(
         IReadOnlyList<PageDetail> pageMap, string blobName)
     {
         const int MaxSectionLength = 1_000;
         const int SentenceSearchLimit = 100;
         const int SectionOverlap = 100;
 
-        var sentenceEndings = new[] { '.', '!', '?' };
-        var wordBreaks = new[] { ',', ';', ':', ' ', '(', ')', '[', ']', '{', '}', '\t', '\n' };
+        var wordBreaks = new[] { ',', '、', ';', ':', ' ', '(', ')', '[', ']', '{', '}', '\t', '\n' };
```

Comment:
Lift this to a class-scoped variable to avoid reallocating the same thing multiple times with each call to `CreateSectionsAsync`.